[wip] codesigning on Darwin

matthewbauer opened the pull request:
This is the start on my quest to get codesign working on macOS. In the end this should close #37838, #18420, #17406 and maybe #15889 and #36223 (lots of other gatekeeper issues especially with jamf and other AVs).

The main thing is packaging "security_systemkeychain" which contains the source for codesign. There is some difficulty here in how Apple has left out a few important packages. We can kind of hack through it with @samdmarshall's OSXPrivateSDK. Probably can't/shouldn't redistribute "codesign" but hopefully still usable in "nativeBuildInputs".

Eventually, I want Hydra to start signing binaries for certain executables. Some that are definitely needed: (it might be useful for others too)

```
- git (for use with credentials-osxkeychain)
- gdb
- darwin.dtrace
- lldb
```

For this to work we will need to run codesign on certain binaries probably in a setup-hook. This is the man for codesign:

```
codesign -s identity [-i identifier] [-r requirements] [-fv] [path ...]
```

We will need an "identity" provided as a config option to Nixpkgs (default to "$CODE_SIGN_IDENTITY"). I'm not sure how we would get the identities on the Hydra machines but I think it would be useful for lots of users (although this would obviously break reproducibility).

To use this right now follow the steps 1-9 here https://gist.github.com/hlissner/898b7dfc0a3b63824a70e15cd0180154 giving the certificate the name "nixpkgs". Now you can use ```darwin.gdb``` and ```darwin.lldb```.

## Usage

You can start using this right now. Configuration is a little tricky so I will take a step by step approach.

1. Open Keychain Access
2. File -> New Keychain....
3. Save keychain in your home directory with the name "nixpkgs".
4. Choose a password that you are okay with leaving plaintext.
5. In the menu, open Keychain Access > Certificate Assistant > Create a certificate
6. Enter name "nixpkgs", Identity type "self-signed root" and Certificate type "Code signing". (Also check "Let me override defaults")
7. Click continue until the "Specify a Location for the Certificate" opens. Select the "nixpkgs" keychain from the drop down.
8. Find the created certificate and right click "Get Info".
9. Under "Trust" select "Always Trust" (this part is optional)
10. Add the following to the file at "$HOME/.config/nixpkgs/config.nix":

```
{
  keychain = {
    file = /Users/<USER>/nixpkgs.keychain;
    identity = "nixpkgs";
    password = "<PASSWORD>";
  };
}
```

matthewbauer committed:
antlr: supports darwin

matthewbauer committed:
xcbuild: add framework product

matthewbauer committed:
xcbuild: set SYMROOT and OBJROOT

matthewbauer committed:
treewide: change install references for xcbuild

now build directory is "Products/Release/"

matthewbauer committed:
pbzx: init at 1.0.2

matthewbauer committed:
swift-corelibs: init

moves swift-corelibs-corefoundation to own dir

also adds libdispatch

matthewbauer committed:
CoreSymbolication: init

matthewbauer committed:
darling: init

matthewbauer committed:
dtrace: add xcode drv

darwin.dtrace-xcode contains the dtrace bin

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=b1e694d4-d88a-4ec5-871e-d0fb3e6f4e52)

Attempted: antlr, pbzx, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
checking for references to /tmp/nix-build-pbzx-1.0.2.drv-0 in /nix/store/rxgb0ci2f86sz0ymvx9572k84vwnjkf9-pbzx-1.0.2...
building path(s) ‘/nix/store/6m62s8qxkic44d9312ymgvym820fvjmm-MacOSX.platform’
building path(s) ‘/nix/store/20q5npx4zchlm4znzmlr86n3c3zkhbkf-xcbuild-wrapper-0.1.2-pre’
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/20q5npx4zchlm4znzmlr86n3c3zkhbkf-xcbuild-wrapper-0.1.2-pre
strip is /nix/store/fzcs0fn6bb04m82frhlb78nc03ny3w55-binutils-2.28.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/20q5npx4zchlm4znzmlr86n3c3zkhbkf-xcbuild-wrapper-0.1.2-pre/bin
patching script interpreter paths in /nix/store/20q5npx4zchlm4znzmlr86n3c3zkhbkf-xcbuild-wrapper-0.1.2-pre
checking for references to /tmp/nix-build-xcbuild-wrapper-0.1.2-pre.drv-0 in /nix/store/20q5npx4zchlm4znzmlr86n3c3zkhbkf-xcbuild-wrapper-0.1.2-pre...
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=a1f910aa-90e0-4c5d-ba6c-b8d9ca626a1a)

Attempted: pbzx, xcbuild

The following builds were skipped because they don't evaluate on aarch64-linux: antlr

<details><summary>Partial log (click to expand)</summary><p>

```
building '/nix/store/r2wqaif8xr61hqqgg22lli2dv395w7md-xcbuild-wrapper-0.1.2-pre.drv'...
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/wbwrwbwilp98wcpz8cnmh2s9fakdwc0i-xcbuild-wrapper-0.1.2-pre
strip is /nix/store/3zq400fri5dv7d30lpxlqm2v9y1iis6j-binutils-2.28.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/wbwrwbwilp98wcpz8cnmh2s9fakdwc0i-xcbuild-wrapper-0.1.2-pre/bin
patching script interpreter paths in /nix/store/wbwrwbwilp98wcpz8cnmh2s9fakdwc0i-xcbuild-wrapper-0.1.2-pre
checking for references to /build in /nix/store/wbwrwbwilp98wcpz8cnmh2s9fakdwc0i-xcbuild-wrapper-0.1.2-pre...
/nix/store/apdz1inay8js5lir4n0iwypvvvxq6nhp-pbzx-1.0.2
/nix/store/wbwrwbwilp98wcpz8cnmh2s9fakdwc0i-xcbuild-wrapper-0.1.2-pre
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=bf2e1edc-1e2c-4ee4-9fd2-781d0bfbfc68)

Attempted: antlr, pbzx, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
/bin/echo "installation done"
installation done
post-installation fixup
strip is /nix/store/0fzpxnsanc02i4jsb1yhchjp4p62b2n3-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/lib  /nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/bin  /nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/sbin
patching script interpreter paths in /nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7
/nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/bin/antlr: interpreter directive changed from "/bin/sh" to "/nix/store/x030a63qdilnv02pkivfjg44pdxsh5km-bash-4.4-p19/bin/sh"
/nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/bin/antlr-config: interpreter directive changed from "/bin/sh" to "/nix/store/x030a63qdilnv02pkivfjg44pdxsh5km-bash-4.4-p19/bin/sh"
/nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/sbin/pyantlr.sh: interpreter directive changed from "/usr/bin/env python" to "/nix/store/fqsrb5whxc3w7cvvk4yw6z8pralx2q5l-python-2.7.14/bin/python"
moving /nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/sbin/* to /nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7/bin
```
</p></details>

matthewbauer committed:
libsecurity_utilities: fixup

matthewbauer committed:
libsecurity_codesigning: init

matthewbauer committed:
security_systemkeychain: init

Fixes #37838

matthewbauer committed:
gdb, lldb: codesign it!

This gives us the codesigned derivation darwin.gdb, darwin.lldb, and
darwin.dtrace.

Fixes #18420
Fixes #17406

matthewbauer committed:
xcbuild: fix wrapper

need $wrapper/usr/bin to exist for some SDK stuff

matthewbauer committed:
SmartCardServices: init

matthewbauer committed:
darwin: fixup eval

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=0d80845e-0db1-4a7f-92e2-cb43d91d2e05)

Attempted: antlr, gdb, lldb, pbzx, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/rxgb0ci2f86sz0ymvx9572k84vwnjkf9-pbzx-1.0.2/bin/pbzx
strip is /nix/store/fzcs0fn6bb04m82frhlb78nc03ny3w55-binutils-2.28.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/rxgb0ci2f86sz0ymvx9572k84vwnjkf9-pbzx-1.0.2/bin
patching script interpreter paths in /nix/store/rxgb0ci2f86sz0ymvx9572k84vwnjkf9-pbzx-1.0.2
checking for references to /build in /nix/store/rxgb0ci2f86sz0ymvx9572k84vwnjkf9-pbzx-1.0.2...
/nix/store/fm1dhr4l0114hlh2izsw71d5r4bvinbc-antlr-2.7.7
/nix/store/4pfpq8k5rfh0ylhr4v0dfqjasyl1kwjb-gdb-8.1
/nix/store/640cc5nngpnb9p90vni74cd6wvc86j90-lldb-5.0.1
/nix/store/rxgb0ci2f86sz0ymvx9572k84vwnjkf9-pbzx-1.0.2
/nix/store/b0rmdvjiknmf2bqr525w9892zpirvn4k-xcbuild-wrapper-0.1.2-pre
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=08a96a29-d54e-423b-8c4a-ba147bd28a6b)

Attempted: antlr, gdb, lldb, pbzx, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
strip is /nix/store/0fzpxnsanc02i4jsb1yhchjp4p62b2n3-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/fxsvbmnr7zd7vbd7h9cm4wzj0z7pa36a-xcbuild-wrapper-0.1.2-pre/bin
patching script interpreter paths in /nix/store/fxsvbmnr7zd7vbd7h9cm4wzj0z7pa36a-xcbuild-wrapper-0.1.2-pre
copying path '/nix/store/4rk52w4kpnivna3xx7mygp8ghac28g2i-lldb-5.0.1' from 'https://cache.nixos.org'...
/nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7
/nix/store/42p7z9i09qxk5d94rd6lwm06wd61l2ap-gdb-8.1
/nix/store/4rk52w4kpnivna3xx7mygp8ghac28g2i-lldb-5.0.1
/nix/store/d33rq2isnn060hhvm1d09kyczrvkiydn-pbzx-1.0.2
/nix/store/fxsvbmnr7zd7vbd7h9cm4wzj0z7pa36a-xcbuild-wrapper-0.1.2-pre
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=ca77f7d9-629d-4c9a-8cc2-5beda72ac0dd)

Attempted: gdb, lldb, pbzx, xcbuild

The following builds were skipped because they don't evaluate on aarch64-linux: antlr

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/w38vaa3rhn0ikq5g2s1dvr7cxp3c09ci-xcbuild-wrapper-0.1.2-pre
strip is /nix/store/3zq400fri5dv7d30lpxlqm2v9y1iis6j-binutils-2.28.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/w38vaa3rhn0ikq5g2s1dvr7cxp3c09ci-xcbuild-wrapper-0.1.2-pre/bin
patching script interpreter paths in /nix/store/w38vaa3rhn0ikq5g2s1dvr7cxp3c09ci-xcbuild-wrapper-0.1.2-pre
checking for references to /build in /nix/store/w38vaa3rhn0ikq5g2s1dvr7cxp3c09ci-xcbuild-wrapper-0.1.2-pre...
/nix/store/9wp8kivp1yp8x9gkzx7r4mry9gksn0ak-gdb-8.1
/nix/store/pi9bxcsx5m2jfgn2q2m6zkvijw7dfsb6-lldb-5.0.1
/nix/store/apdz1inay8js5lir4n0iwypvvvxq6nhp-pbzx-1.0.2
/nix/store/w38vaa3rhn0ikq5g2s1dvr7cxp3c09ci-xcbuild-wrapper-0.1.2-pre
```
</p></details>

Mic92 commented:
Without code signing applications are not allowed to use the `ptrace()` system call?

LnL7 commented:
@Mic92 yes and SIP adds even more restrictions.

matthewbauer commented:
I'm having trouble at dealing with macOS keychains in multi-user mode. Single-user mode would probably just work out of the box though.

So we probably will need to import the private key into the builder (might be unsafe given NixOS/nix#8). Anyway this is kind of blocked on #38633.

matthewbauer commented:
@GrahacOfBorg build darwin.gdb

matthewbauer commented:
@GrahamcOfBorg build darwin.gdb

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=046c836e-4505-4bc5-ba1c-ae015d48f63f)

The following builds were skipped because they don't evaluate on x86_64-linux: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowBroken = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=0f676861-64cd-438c-8322-bca66497c0f5)

The following builds were skipped because they don't evaluate on aarch64-linux: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowBroken = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

copumpkin commented:
Hmm, will have to look carefully at this. I didn't think it was really possible to do codesigning in a principled/secure manner in Nix so I'll be happy to be proven wrong, but am also a bit apprehensive 😄 unless you see this purely as a "we give Hydra signing keys and you only get signatures if you get your binaries from the binary cache" deal.

LnL7 reviewed (commented):
This is great! But I'm not sure if it makes sense to sign binaries from inside a build. It's an impurity and allowing any untrusted user to sign binaries doesn't sound safe to me (if that's even possible).

matthewbauer committed:
dtrace: patch impurities

matthewbauer committed:
all-packages: add codesigning!

matthewbauer committed:
darwin.codesign: cleanup

matthewbauer committed:
libsecurity_apple_x509: add more symbols

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=01f301c9-7435-43a3-b5a6-d9d5dc84918a)

Attempted: antlr, gdb, lldb, pbzx, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/fm1dhr4l0114hlh2izsw71d5r4bvinbc-antlr-2.7.7
/nix/store/4pfpq8k5rfh0ylhr4v0dfqjasyl1kwjb-gdb-8.1
/nix/store/640cc5nngpnb9p90vni74cd6wvc86j90-lldb-5.0.1
/nix/store/rxgb0ci2f86sz0ymvx9572k84vwnjkf9-pbzx-1.0.2
/nix/store/b0rmdvjiknmf2bqr525w9892zpirvn4k-xcbuild-wrapper-0.1.2-pre
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=49a9c036-ca31-4b31-9f4d-6d638b2696e3)

Attempted: gdb, lldb, pbzx, xcbuild

The following builds were skipped because they don't evaluate on aarch64-linux: antlr

<details><summary>Partial log (click to expand)</summary><p>

```

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/9wp8kivp1yp8x9gkzx7r4mry9gksn0ak-gdb-8.1
/nix/store/pi9bxcsx5m2jfgn2q2m6zkvijw7dfsb6-lldb-5.0.1
/nix/store/apdz1inay8js5lir4n0iwypvvvxq6nhp-pbzx-1.0.2
/nix/store/w38vaa3rhn0ikq5g2s1dvr7cxp3c09ci-xcbuild-wrapper-0.1.2-pre
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=77f1e2b4-b63f-4b93-8b75-11d7c406ad72)

Attempted: antlr, gdb, lldb, pbzx, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
strip is /nix/store/0fzpxnsanc02i4jsb1yhchjp4p62b2n3-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/d33rq2isnn060hhvm1d09kyczrvkiydn-pbzx-1.0.2/bin
patching script interpreter paths in /nix/store/d33rq2isnn060hhvm1d09kyczrvkiydn-pbzx-1.0.2
copying path '/nix/store/4rk52w4kpnivna3xx7mygp8ghac28g2i-lldb-5.0.1' from 'https://cache.nixos.org'...
/nix/store/cp4m9h5r7y0npzgsizah0q4p94k5xmix-antlr-2.7.7
/nix/store/42p7z9i09qxk5d94rd6lwm06wd61l2ap-gdb-8.1
/nix/store/4rk52w4kpnivna3xx7mygp8ghac28g2i-lldb-5.0.1
/nix/store/d33rq2isnn060hhvm1d09kyczrvkiydn-pbzx-1.0.2
/nix/store/fxsvbmnr7zd7vbd7h9cm4wzj0z7pa36a-xcbuild-wrapper-0.1.2-pre
```
</p></details>

matthewbauer commented:
@GrahamcOfBorg build darwin.gdb

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=08214fb6-f32a-4702-a571-c6eab129fc6d)

The following builds were skipped because they don't evaluate on x86_64-darwin: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `darwin.gdb' because:
[31;1merror:[0m attribute 'gdb' in selection path 'darwin.gdb' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=662155c9-f791-44f6-b1e9-19f483e3a429)

The following builds were skipped because they don't evaluate on x86_64-linux: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `darwin.gdb' because:
error: attribute 'gdb' in selection path 'darwin.gdb' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=7321e76e-ba4c-463b-b6de-ecbab62fbdb2)

The following builds were skipped because they don't evaluate on aarch64-linux: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `darwin.gdb' because:
[31;1merror:[0m attribute 'gdb' in selection path 'darwin.gdb' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=40804292-4211-4413-a87e-c8222d535cf0)

The following builds were skipped because they don't evaluate on x86_64-darwin: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `darwin.gdb' because:
[31;1merror:[0m attribute 'gdb' in selection path 'darwin.gdb' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=cc3ce37c-4dcc-4452-a4dc-7ff651054ef9)

The following builds were skipped because they don't evaluate on x86_64-linux: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `darwin.gdb' because:
[31;1merror:[0m attribute 'gdb' in selection path 'darwin.gdb' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=573d427b-f837-4e56-b456-5f965999c201)

The following builds were skipped because they don't evaluate on aarch64-linux: darwin.gdb

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `darwin.gdb' because:
[31;1merror:[0m attribute 'gdb' in selection path 'darwin.gdb' not found

```
</p></details>

matthewbauer commented on pkgs/os-specific/darwin/apple-source-releases/default.nix:
```diff
@@ -201,7 +205,8 @@ let
     Csu             = applePackage "Csu"               "osx-10.11.6"     "0yh5mslyx28xzpv8qww14infkylvc1ssi57imhi471fs91sisagj" {};
-    dtrace          = applePackage "dtrace"            "osx-10.11.6"     "0pp5x8dgvzmg9vvg32hpy2brm17dpmbwrcr4prsmdmfvd4767wc0" {};
+    dtrace          = applePackage "dtrace"            "osx-10.11.6"     "04mi0jy8gy0w59rk9i9dqznysv6fzz1v5mq779s41cp308yi0h1c" {};
+    dtrace-xcode    = applePackage "dtrace/xcode.nix"  "osx-10.11.6"     "04mi0jy8gy0w59rk9i9dqznysv6fzz1v5mq779s41cp308yi0h1c" {};
```
Ideally we would merge these into one- but we run into an issue in bootstrapping.

matthewbauer commented:
> But I'm not sure if it makes sense to sign binaries from inside a build. It's an impurity

So what I'm hoping is to create a kind of wrapper for them. That way you can still get the "pure" derivation. 

>  and allowing any untrusted user to sign binaries doesn't sound safe to me (if that's even possible)

Well it will go in the user's config option. So the untrusted user will already need to have access to a keychain in the first place. I guess an open question will be whether Nix can be tricked into giving up the "config" options (hopefully it stays private).

copumpkin reviewed (commented):
I've mentioned this in some other threads (I swear I created an issue for tracking it but I can't find it now), but I do think code signing needs to be a very carefully guarded option. I think of it like a fancier version of setuid and the pain we've had to deal with in the past around it. But unlike setuid, the mere presence of a valid signing key on a machine is basically circumventing Apple's entire "signed code" checking mechanism, because if an attacker gains control over that key (and local codesigning checks are trying to protect even against local attackers), they can sign arbitrary code.

Furthermore, even if we try to be exceptionally careful about it, the `security` interface to the keychain is pretty footgun-flavored. By that, I mean that Apple's keychain services authenticate/authorize access by the hash (as far as I've been able to tell) of the program making the request. So if you make a request from the CLI to get at a private key with `security`, the UI prompt will ask if the user wants to grant `security` that access. If the user says to always authorize it, subsequent requests for that key using `security` will succeed.

This means that the keychain API is really only safe to use from compiled code. Python has an interface to it, but it suffers from a similar issue, and you'll be granting access to `python` rather than the specific script.

Now, given that `codesign` is its own program I don't know how Apple deals with authorization to access it, but I do think we need to be quite careful and deliberate about how we use it.

matthewbauer commented on pkgs/top-level/darwin-packages.nix:
```diff
@@ -57,12 +57,46 @@ in
+    mkdir -p $PWD/Library/Keychains
+    cp ${config.keychain.file} $PWD/Library/Keychains
+    KEYCHAIN=$(basename $PWD/Library/Keychains/*)
+    security unlock-keychain -p $PASS $KEYCHAIN
```
So this is definitely a hacky demo - and not ready to be "trusted" by the average Mac user - but I think it can be made to work.

I think this should actually work in a sandbox. There are a few hardcoded places that security will look for "Keychains" but I think it would just skip them if it was in a sandbox. Since we are passing the path here, it should be okay. One question I'm not sure of though is how to get the "substituter" to work with different signatures, identities, etc.

I'm not sure if we can get much out of doing this in Nix. The binary caches already have signatures so that means that everything in the Nix store should already be "trusted" by the user- at least as much as you trust Nixpkgs. "Codesigning" in Nixpkgs is really just for Gatekeeper and other macOS stuff. Putting it all in Nixpkgs is definitely hacky but I think it's actually the best way to go.

LnL7 replied:
I'm not sure, doesn't giving access to the keychain in a multi-user install mean that the signing key is essentially public. If that's the case it could be used to sign any arbitrary binary, what does it mean to trust the key then?

This would still be true for hydra builds, even if this is implemented in nix. Except that it's guaranteed to be done by the hydra infrastructure.

matthewbauer commented:
So this is turning out to be more difficult than I thought. I think I will open a separate PR with the "good" commits from this (leaving out any actual codesigning) that help to get codesigning working. Most likely will  be: e51405db3e5d5e89e1a9dd8251ef00df0f771913...77ac9c50dce15b681d368b508d88b8b6c8440d15

LnL7 commented:
Yeah, we should definitively include that.

matthewbauer committed:
wip

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=1487e232-1bea-4d7f-9da5-16b9718787b9)

Attempted: gdb, lldb, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
  /nix/store/yr20falfjsiczglc1y6bi8jm2shmd3bf-llvm-5.0.1-lib
copying path '/nix/store/yr20falfjsiczglc1y6bi8jm2shmd3bf-llvm-5.0.1-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/h1wzgamdwdphbgk4mxxnkn82inbsxkzz-xcbuild-0.1.2-pre' from 'https://cache.nixos.org'...
copying path '/nix/store/0crd5c951217xbgsdm1jkpqzbq69mlr8-gperf-3.1' from 'https://cache.nixos.org'...
copying path '/nix/store/00bhyvb34bym77q6vyy4i8k3izjppma4-nixpkgs.xctoolchain' from 'https://cache.nixos.org'...
copying path '/nix/store/mn2lqfwq6qvfz0nhy7vrsm7zg6q8dlnq-xcbuild-wrapper-0.1.2-pre' from 'https://cache.nixos.org'...
copying path '/nix/store/y02cf16x2wr7kagbyijbkx815ndazy4b-lldb-5.0.1' from 'https://cache.nixos.org'...
/nix/store/q2w8mjqwi0ycira8nm09xrdgqqs8yp50-gdb-8.1
/nix/store/y02cf16x2wr7kagbyijbkx815ndazy4b-lldb-5.0.1
/nix/store/mn2lqfwq6qvfz0nhy7vrsm7zg6q8dlnq-xcbuild-wrapper-0.1.2-pre
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38624&attempt_id=5478b171-8f2a-4e22-a6dd-cd95fc578d52)

Attempted: gdb, lldb, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/34ll1dzs52k1gw8d44ivw3ry9hlb8pk1-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/65vlr4sb6rz125ia3krqhwh29hv8416x-lldb-5.0.1' from 'https://cache.nixos.org'...
copying path '/nix/store/j8wyhwi347990b3rcd5mf0iqr1ng6jd1-binutils-wrapper-2.30' from 'https://cache.nixos.org'...
copying path '/nix/store/7ld1q91jn1wf37g9r1y2diiqq3gz33n0-clang-5.0.1' from 'https://cache.nixos.org'...
copying path '/nix/store/sli8mhk4ybz76dsq27s0ar0jvd0501lf-clang-wrapper-5.0.1' from 'https://cache.nixos.org'...
copying path '/nix/store/vyr7ji9pd56k1cn03vdd30nhaqq48132-nixpkgs.xctoolchain' from 'https://cache.nixos.org'...
copying path '/nix/store/9g9mag7w2qmljc73xzrjh7zq2a7qhphg-xcbuild-wrapper-0.1.2-pre' from 'https://cache.nixos.org'...
/nix/store/hw6j3lkcsdlmc9vr00qj0d8771ikjkh6-gdb-8.1
/nix/store/65vlr4sb6rz125ia3krqhwh29hv8416x-lldb-5.0.1
/nix/store/9g9mag7w2qmljc73xzrjh7zq2a7qhphg-xcbuild-wrapper-0.1.2-pre
```
</p></details>

matthewbauer commented:
I think the big problem with this is we can't really store secrets in the Nix store. The key will be publicly available to anyone who can access the .drv file. I don't think it would be a good idea to do this on Hydra until Nix can support some sort of private key impurity.

ElvishJerricco commented:
Hm... It's too bad there's not a way to tell macOS that a binary should be trusted without altering it. Currently, if we wanted to sign binaries from Nix, it would be impure and it would produce binaries that can't be `nix copy`'d over to other machines. If macOS just had a DB of signatures separate from the binaries themselves, then we could have unaltered builds and just add their signatures to such a DB after the fact with a Nix plugin.

ElvishJerricco commented:
Welp, I guess I should have read the manual before commenting, because it appears that exactly that is possible:

```
     -D, --detached filename
             When signing, designates that a detached signature should be written to the specified file. The code being
             signed is not modified and need not be writable.  When verifying, designates a file containing a detached sig-
             nature to be used for verification. Any embedded signature in the code is ignored.

     --detached-database
             When signing, specifies that a detached signature should be generated as with the --detached option, but that
             the resulting signature should be written into a system database, from where it is made automatically available
             whenever apparently unsigned code is validated on the system.
             Writing to this system database requires elevated process privileges that are not available to ordinary users.
```

ElvishJerricco commented:
Just tested. Seems to partially work:

```
$ nix-env -f "<nixpkgs>" -iA gdb
$ sudo codesign -s "$signing-identity" --detached-database $(readlink $(which gdb))
$ codesign --verify $(readlink $(which gdb)) && echo success
success
```

However, GDB still does not work... Still suggests that it's not codesigned. So for some reason, `codesign --verify` is not quite the same thing. Anyone have any thoughts?

ElvishJerricco commented:
@copumpkin Ah. Also, I'm guessing all the dynamic libraries `gdb` depends on need to be signed.

copumpkin commented:
Yup

ElvishJerricco commented:
Ah ha! Got it working. Had to sign all the dylibs it used, and I had to open the "Get Info" page for the certificate, open the "Trust" section, and set "Code Signing" to "Always Trust". Once I'd done that, I could use `gdb` normally.

Does this seem like a viable path forward? Simply making zero changes in nixpkgs and instead finding some way to automate the detached signing of binaries?

copumpkin commented:
@ElvishJerricco I think the detached signatures will be helpful but still don't address most of what I wrote in my longer comment above, which I don't think is really easily addressable. We could have a special case to cover the common case and have Hydra sign a handful of binaries in some ad-hoc way, which would get most of us a working gdb/lldb/dtrace/git, but the local signing story is still pretty unclear to me, fundamentally. We can work around it but it would basically be negating the value of one security control Apple deliberately put in place, and I don't want to be known as "the package manager that makes your computer less secure" unless we give people a very loud option for opting into it.

ElvishJerricco commented:
@copumpkin I don't think we need Hydra to provide macOS signatures in any way. I see two possible paths forward, both using local signing:

- Provide automated tooling for letting users do detached signing of paths manually. A Nix command, or something, that adds detached signatures for all the binaries in the runtime closure of a path.
- Add a `macos-signing-identity` option to Nix. Whenever a store path is realized, iff it has a trusted Nix signature, automatically perform the signing described above. The issue here is revoking Nix keys; you'd want to keep each Nix key in correspondence with a codesigning key, so they can be revoked in tandem.

The issue is that it'll prompt the user for every binary we try to sign, but we can maybe work around that with with some kind of TTL / caching.

ElvishJerricco commented:
Hydra could in theory also provide detached macOS signatures, if there's a way to add a pre-computed detached signature to this magic database. Maybe Nix should have a more pluggable concept of metadata, so that we can ship arbitrary kinds of signatures with a path, and just use Nix plugins to verify them properly?

ElvishJerricco commented:
Then it sounds like option 1 would be ok. Nix does not have to sign anything; users can just sign things themselves when they want. One way or another, their gdb has to be signed for it to work, and that signature isn't going to come from apple if gdb comes from nix. So sure, make it something they have to do themselves. But maybe automate it so they can do so easily.

matthewbauer commented:
This doesn't on its own effect Apple's security. It just gives you a way of knowing the identity of who built your binary. It will all depend on what certificate you use to sign things. If its locally generated, then you will need to manually "trust" that certificate. Otherwise you can use a real one from the developer program.

The main issue is that you can't really keep a file a secret in Nix. I would be interested if anyone has ideas for alternatives to this. Your private key should never be made publicly available through the Nix store.

ElvishJerricco commented:
@matthewbauer I would suggest that you shouldn't sign binaries within a Nix expression. That should be left to external tooling, like `codesign --detached-database`
